typing/ident.ml

```diff
@@ -23,9 +23,12 @@ type t =
   | Scoped of { name: string; stamp: int; scope: int }
   | Global of string
   | Predef of { name: string; stamp: int }
+  | Unscoped of { name: string; stamp: int }
```

Comment:
There is no need for a new `Unscoped` constructor, the existing `Local` constructor already exists for representing bound identifiers. I expect that the places where you think that `Unscoped` should not appear should also not have `Local` either. The only place where I know that they do appear unexpectedly is the root of [an existing bug](https://github.com/ocaml/ocaml/issues/9771).

@trefis suggestion to use `Scoped` with an infinite level is also reasonable, and consistent with how we deal with bound type variables. @trefis and I have been discussing how we might improve OCaml's representation of bound type variables. Taking advantage of these improvements with module explicits would require using a `Scoped` with an infinite level, so I suspect that is why he would prefer to do it this way.

Reply:
I've gone back and forth on this, but on balance I think it's best to keep this as a dedicated constructor.

Concretely,
* the typechecker won't always do 'the right thing' when there's a possible escape, since the level of the type can be lower than the target level, even though the identifier's level would be at/around the highest level
  - in essence, giving this an artificially high scope violates the decreasing levels assumption, so the optimisations using that will cause the check to be skipped
  - it doesn't seem great to have 2 distinct codepaths checking for and reporting escapes differently, depending arbitrarily (from the user's perspective) on whether the type containing the escaping path is at a higher level than the variable or not
* the checks for escaping identifiers are easy and cheap at the moment, because the `Ident.is_unscoped` function (and friends) can tell immediately whether the identifier is valid or not
  - I worry about applying a similar approach with `Local` identifiers, since I'm not entirely clear on how or where they might turn up
  - using a dedicated high level will require carefully adding special handling for it in more places
* giving the `Unscoped` constructor a level has the potential to hide any logic errors, by allowing them to leak into the environment etc. instead of crashing the compiler
* `Unscoped` identifiers shouldn't appear in the environment, but the compiler currently depends on placing `Local` identifiers in the environment
  - as a test, returning an `Unscoped` from `Ident.create_local` results in a crash from this early on in compiling the standard library.

If the proposed levels system will make handling these identifiers more elegant, perhaps it would make sense to give them a level when implementing that instead?